test/parallel/test-async-local-storage-deep-stack.js

```diff
@@ -0,0 +1,15 @@
+'use strict';
+const common = require('../common');
+const { AsyncLocalStorage } = require("async_hooks");
+
+// Regression test for: https://github.com/nodejs/node/issues/34556
+
+const als = new AsyncLocalStorage();
+
+const done = common.mustCall();
+
+function run(count) {
+    if (count !== 0) return als.run({}, run, --count)
+    done();
+}
+run(1000);
```

Comment:
Fwiw, this still crashes for me for larger values (e.g. 2000), but with a different error and unrelated to the bug that this is fixing, and depending on `--stack-size`, i.e. there’s probably something wrong with our error handling here.